sw/device/silicon_creator/lib/drivers/keymgr_functest.c

```diff
@@ -0,0 +1,175 @@
+ *
+ * @param page_id Info page ID to write to.
+ * @param data Data to write.
+ * @param size Number of 4B words to write from `data` buffer.
+ */
+static void write_info_page(uint32_t page_id, const uint32_t *data,
+                            size_t size) {
+  mp_region_t info_region = {.num = page_id,
+                             .base = 0x0,  // only used to calculate bank id.
+                             .size = 0x1,  // unused for info pages.
+                             .part = kInfoPartition,
+                             .rd_en = true,
+                             .prog_en = true,
+                             .erase_en = true,
+                             .scramble_en = false};
+  flash_cfg_region(&info_region);
+
+  uint32_t address = FLASH_MEM_BASE_ADDR + page_id * flash_get_page_size();
+
+  ASSERT_EQZ(flash_page_erase(address, kInfoPartition));
+  ASSERT_EQZ(flash_write(address, kInfoPartition, data, size));
+
+  for (size_t i = 0; i < size; ++i) {
+    uint32_t got_data;
+    ASSERT_EQZ(flash_read(address + i * sizeof(uint32_t), kInfoPartition,
+                          /*size=*/1, &got_data));
+    CHECK(got_data == data[i]);
+  }
+  info_region.rd_en = false;
+  info_region.prog_en = false;
+  info_region.erase_en = false;
+  flash_cfg_region(&info_region);
+}
+
+static void init_flash(void) {
+  // setup default access for data partition
+  flash_default_region_access(/*rd_en=*/true, /*prog_en=*/true,
+                              /*erase_en=*/true);
+}
+
+/** Key manager configuration steps performed in mask ROM. */
+rom_error_t keymgr_rom_test(void) {
+  ASSERT_OK(keymgr_check_state(kKeyMgrStateReset));
+  keymgr_set_next_stage_inputs(&kBindingValueRomExt, kMaxVerRomExt);
+  return kErrorOk;
+}
+
+/** Key manager configuration steps performed in ROM_EXT. */
+rom_error_t keymgr_rom_ext_test(void) {
+  ASSERT_OK(keymgr_check_state(kKeyMgrStateReset));
+
+  const uint16_t kEntropyReseedInterval = 0x1234;
+  ASSERT_OK(keymgr_init(kEntropyReseedInterval));
+  keymgr_advance_state();
+  ASSERT_OK(keymgr_check_state(kKeyMgrStateInit));
+
+  // FIXME: Check `kBindingValueRomExt` before advancing state.
+  keymgr_advance_state();
+  ASSERT_OK(keymgr_check_state(kKeyMgrStateCreatorRootKey));
+
+  keymgr_set_next_stage_inputs(&kBindingValueBl0, kMaxVerBl0);
+
+  // FIXME: Check `kBindingValueBl0` before advancing state.
+  keymgr_advance_state();
+  ASSERT_OK(keymgr_check_state(kKeyMgrStateOwnerIntermediateKey));
+  return kErrorOk;
+}
+
+bool test_main(void) {
+  rom_error_t result = kErrorOk;
+
+  lifecycle_state_t lc_state = lifecycle_state_get();
+  LOG_INFO("lifecycle state: %s", lifecycle_state_name[lc_state]);
+
+  // Initialize flash secrets.
+  init_flash();
+  write_info_page(kFlashInfoPageIdCreatorSecret, kCreatorSecret,
+                  ARRAYSIZE(kCreatorSecret));
+  write_info_page(kFlashInfoPageIdOwnerSecret, kOwnerSecret,
+                  ARRAYSIZE(kOwnerSecret));
```

Comment:
nit: maybe these should be `init_flash()`.

Reply:
Done.